telemetry/src/lib.rs

```diff
@@ -0,0 +1,235 @@
+use std::{collections::VecDeque, sync, sync::Arc, time::Duration};
+
+use observability_deps::tracing::error;
+use serde::Serialize;
+use tokio::sync::mpsc;
+
+/// This is the core type exposed from this crate which can be
+/// cloned in different places to send the `InternalTelemetryMessage`.
+#[derive(Clone)]
+pub struct TelemetryHandle {
+    /// mpsc sender, safe to clone
+    sender: mpsc::Sender<InternalTelemetryMessage>,
+}
+
+impl TelemetryHandle {
+    pub async fn new(
+        instance_id: String,
+        os: String,
+        influx_version: String,
+        storage_type: String,
+        cores: u32,
+    ) -> Self {
+        let store = TelemetryStore::new(instance_id, os, influx_version, storage_type, cores);
+        // TODO: decide the buffer size
+        let (sender, receiver) = mpsc::channel(100);
+        let telem_msg_handler = TelemetryMesageHandler::new(receiver, store.clone());
+        telem_msg_handler.handle_messages().await;
+        send_telemetry_in_background(store.clone()).await;
+        TelemetryHandle { sender }
+    }
+
+    pub async fn send(&self, msg: InternalTelemetryMessage) {
+        // TODO: handle failures in send
+        let _ = self.sender.send(msg).await;
+    }
+}
+
+/// The message types that can be sent through the `TelemetryHandle`.
+/// Currently we only need to collect numerical data periodically
+/// so just sticking with u64 as associated data is good enough
+pub enum InternalTelemetryMessage {
+    Writes(u64),
+    Reads(u64),
+    CpuUsage(u64),
+    CpuUtilization(u64),
+}
+
+/// This is responsible for receiving messages from `TelemetryHandle`
+struct TelemetryMesageHandler {
+    receiver: mpsc::Receiver<InternalTelemetryMessage>,
+    store: TelemetryStore,
+}
+
+impl TelemetryMesageHandler {
+    pub fn new(receiver: mpsc::Receiver<InternalTelemetryMessage>, store: TelemetryStore) -> Self {
+        TelemetryMesageHandler { receiver, store }
+    }
+
+    pub async fn handle_messages(mut self) {
+        tokio::spawn(async move {
+            loop {
+                if let Some(msg) = self.receiver.recv().await {
+                    // do something with msg
+                    self.store.handle_new_message(msg);
+                }
+            }
+        });
+    }
+}
+
+/// All of the internal messages are collected in this store
+#[derive(Clone)]
+struct TelemetryStore {
+    inner: Arc<sync::Mutex<TelemetryStoreInner>>,
```

Comment:
I would expect the `TelemetryStore` to be the thing that is Arc wrapped with copies of that getting sent around. Then you can have its interface clean with whatever methods you need, which lock the internal Mutex.

Also, I'd use the Mutex from `parking_lot`